examples/ex-sliding.cpp

```diff
@@ -0,0 +1,523 @@
+    to move along a line (eg the intersection of two normal-constrained
+    planes), and a three-row constraint is fully constrained (equivalent
+    to MFEM's usual essential boundary conditions). The lagrange_rowstarts
+    array is filled in to describe the structure of these constraints.
+
+    When two attributes intersect, this version will combine constraints,
+    so in 2D the point at the intersection is fully constrained (ie,
+    fixed in both directions). This is the wrong thing to do if the
+    two boundaries are (close to) parallel at that point.
+
+    @param[in] fespace              A vector finite element space
+    @param[in] constrained_att      Boundary attributes to constrain
+    @param[out] lagrange_rowstarts  The rowstarts for separately
+                                    eliminated constraints, possible
+                                    input to EliminationCGSolver
+
+    @return a constraint matrix
+
+    @todo use FiniteElementSpace instead of ParFiniteElementSpace, but
+    we need tdofs in parallel case. */
+SparseMatrix * BuildNormalConstraints(ParFiniteElementSpace& fespace,
+                                      Array<int>& constrained_att,
+                                      Array<int>& lagrange_rowstarts)
+{
+   int dim = fespace.GetVDim();
+
+   // mapping from dofs (colums of constraint matrix) to constraints
+   // (rows of constraint matrix)
+   // the indexing is by tdof, but we only bother with one tdof per node
+   std::map<int, int> dof_constraint;
+   // constraints[j] is a map from attribute to row number
+   std::vector<std::map<int, int> > constraints;
+   int n_constraints = 0;
+   int n_rows = 0;
+   for (int att : constrained_att)
+   {
+      std::set<int> constrained_tdofs;
+      for (int i = 0; i < fespace.GetNBE(); ++i)
+      {
+         if (fespace.GetBdrAttribute(i) == att)
+         {
+            Array<int> dofs;
+            fespace.GetBdrElementDofs(i, dofs);
+            for (auto k : dofs)
+            {
+               int vdof = fespace.DofToVDof(k, 0);
+               int tdof = fespace.GetLocalTDofNumber(vdof);
+               if (tdof >= 0) { constrained_tdofs.insert(tdof); }
+            }
+         }
+      }
+      for (auto k : constrained_tdofs)
+      {
+         auto it = dof_constraint.find(k);
+         if (it == dof_constraint.end())
+         {
+            dof_constraint[k] = n_constraints++;
+            constraints.emplace_back();
+            constraints.back()[att] = n_rows++;
+         }
+         else
+         {
+            constraints[it->second][att] = n_rows++;
+         }
+      }
+   }
+
+   // reorder so constraints eliminated together are grouped
+   // together in rows
+   // (this is perhaps not the best way to organize things)
+   std::map<int, int> reorder_rows;
+   int new_row = 0;
+   MFEM_VERIFY(lagrange_rowstarts.Size() == 0,
+               "rowstarts must be empty to start!");
```

Comment:
If people are doing this process multiple times let's say for changing BCs over time you could maybe use the `Array::DeleteAll()` function to enforce this.